### PR TITLE
chore: Use span fields as metrics labels

### LINKED ISSUE #2660: Include internal trace context with internal metrics
https://github.com/timberio/vector/pull/1953 introduced a new `internal_metrics` source for Vector's internal metrics. Metrics are defined within events, you can see [an example here](https://github.com/timberio/vector/blob/4a5c201428e20e4ab3aed7f88e7e124124a64719/src/internal_events/elasticsearch.rs#L40-L44). A caveat noted in https://github.com/timberio/vector/pull/1953 is the ability to carry over the internal tracing context:

> A big benefit of `tracing` is its ability to propagate contextual information about where you are in the system. We use this extensively for our logging to tell you what type and the name of the current component without having to manually pass that data around everywhere just for log messages. This same functionality would be very beneficial for naming and tagging metric data, so a tracing based solution is what we started with and what I spent quite a bit of time exploring.
>
> Unfortunately, the state of using `tracing` to collect metrics is extremely early right now. What we want to do is definitely possible, but both the exact implementation and the public API are still undergoing a lot of experimentation. This seems like an area we could jump in and help advance the state of the art, but my current read is that we need a metrics system sooner than that work will be ready, even with our help.

This issue represents the work to solve this.

### PR DESCRIPTION
**Pull request by MOZGIII**
Closes #2660

Commits are semantic, should be easy to follow along through history.

### CODE REVIEW

File: src/internal_events/lua.rs
Comment (Hoverbear) [inline src/internal_events/lua.rs:23]: Why are all these integers turning into floats?

File: src/metrics.rs
Comment (Hoverbear) [inline src/metrics.rs:30]: Ohhhh okay cool!

File: Cargo.toml
Comment (Hoverbear) [inline Cargo.toml:69]: Can you donate these repos to the timberio org before merge? 😁

File: src/internal_events/lua.rs
Comment (MOZGIII) [reply src/internal_events/lua.rs:23]: A change in `metrics` crate. I don't think it matters for our use cases here.

File: Cargo.toml
Comment (MOZGIII) [reply Cargo.toml:69]: I hope to merge the PR to upstream before the merge. I expect this to happen today or tomorrow.

Comment (Hoverbear) [reply Cargo.toml:69]: Okay perfect!!!!

**Review by Hoverbear — APPROVED** [review]
This overall

**Comment by MOZGIII** [comment]
https://github.com/metrics-rs/metrics/pull/87 has been merged, will switch the crate at `Cargo.toml` after new `0.13-alpha`s ship!

**Comment by MOZGIII** [comment]
I found out that the notation we're using at topology at `spawn_sink`, `spawn_transform` and `spawn_source` injects the `name` and `type` as span data. I presume this is what we want to capture. One downside there is that names are a bit too generic, and can be easily overwritten in inner spans or logging calls.
I suppose this on it's own is not a good idea, and we should probably name them a bit more distinctively.

Also, the code I have filters by `component_name` annotation, but we can easily add more fields. We can't check the span name currently (i.e. to only grab the data from the `sink`/`transform`/`source`) - it's doable, but it doesn't seem like a good approach in the first place because it doesn't help with potential overrides at logging calls caused by inner spans or literal arguments at callsite. We probably want to just solve all those at once via specifying different span field names.

**Comment by binarylogic** [comment]
I don't mind either way, but I would choose `component_id` and `component_type` for the context names.

File: src/kubernetes/state/instrumenting.rs
Comment (bruceg) [inline src/kubernetes/state/instrumenting.rs:103]: Would it make sense to roll this into `fn get_controller`?

**Review by bruceg — APPROVED** [review]
Looks okay, though the title really underplays what is going on here as it looks like a significant rework of our internal metrics underlay.

File: src/kubernetes/state/instrumenting.rs
Comment (MOZGIII) [reply src/kubernetes/state/instrumenting.rs:103]: I thought about it, and it looks like it only makes sense for tests. Otherwise, to extend this idea further, we could use [`once_cell::sync::Lazy`](https://docs.rs/once_cell/1.4.1/once_cell/sync/struct.Lazy.html) for the initialization, and get rid of the error altogether.
However, my thinking is, in the usual application flow, we'd rather have more control on the initialization time of the metrics, rather than doing it lazily. The problem there is if we couple `get_controller` with `init` like that, then people might assume it's ok to just use `get_controller` and skip the `init` in a proper location. But `init` is required, because it also makes `metrics` macros work, and those can't be coupled with `init` internally. Thus, it's easier for us to reason about this whole system if we don't couple `get_controller` with `init`.

Actually, I probably should call `init` here unconditionally to avoid the confusion. It's implemented as an idempotent call, so it'll be ok.

**Comment by MOZGIII** [comment]
> I don't mind either way, but I would choose `component_id` and `component_type` for the context names.

Ended up using `topology_component_name` and `topology_component_type`, because the `component_type` is already used as a literal value at the the `emit!`s.

Did a test run - it works great in practice. I suggest interested parties take the final look. I'll probably merge this on Monday.

**Comment by binarylogic** [comment]
Yeah, I'd prefer that we use `component_name` and `component_type` if possible. Can we not fix that? Not a huge deal, but

**Comment by MOZGIII** [comment]
@binarylogic I’d prefer that we don’t define the same field in two locations to eliminate the confusion with where they come from. So, I suggest removing the literals `component_type` from the `emit!` invocations from each of the components.
Alternatively, we can remove the `component_type` from the toolpology.

Not removing any would lead to metrics having two `component_type` labels with potentially different values (in the case where source is composed using other source/transform for instance).

@lukesteensen what do you think?

**Comment by binarylogic** [comment]
It's fine. You've got 2 approvals, let's merge.

**Comment by lukesteensen** [comment]
> I suggest removing the literals `component_type` from the `emit!` invocations from each of the components.

Yes, my assumption was that we'd remove the manual tags from the event definitions once we're able to derive them from the span data. This is especially important for events that are emitted from shared components, where we previously couldn't know the exact type statically.

**Comment by MOZGIII** [comment]
I'll create a follow-up PR to put this in order.

**Pull request merged**
